File: day3.rb
file = File.open("day3.txt")
data = file.readlines.map(&:chomp)

gamma = ""
epsilon = ""

(0..data.length - 1).each do | i |
    data[i] = data[i].split("").map(&:to_i)
end

columns = data.transpose()

columns.each do | i |
    if i.sum > i.length / 2
        gamma += "1"
        epsilon += "0"
    else
        gamma += "0"
        epsilon += "1"
    end
end

puts "Part one: #{gamma.to_i(2) * epsilon.to_i(2)}"

co_data =* data

(0..co_data[0].length - 1).each do | i |
    temp = Array.new
    columns = co_data.transpose()
    if columns[i].sum == columns[i].length / 2
        a = 1
    elsif columns[i].sum > columns[i].length / 2
        a = 1
    else
        a = 0
    end
    co_data.each do | j |
        if j[i] == a
            temp << j
        end
    end
    co_data =* temp
    if co_data.length == 1
        $c = co_data[0].join("")
    end
end

ox_data =* data

(0..ox_data[0].length - 1).each do | i |
    temp = Array.new
    columns = ox_data.transpose()
    if columns[i].sum == columns[i].length / 2
        a = 0
    elsif columns[i].sum > columns[i].length / 2
        a = 0
    else
        a = 1
    end
    ox_data.each do | j |
        if j[i] == a
            temp << j
        end
    end
    ox_data =* temp
    if ox_data.length == 1
        $d = ox_data[0].join("")
    end
end

puts "Part two: #{$c.to_i(2) * $d.to_i(2)}"

File: day3.py
with open("day3.txt") as f:
    data = [x.strip() for x in f]

gamma = list()
epsilon = list()

def popular_bit(x):
    if sum(x) == (len(x) / 2):
        return "both"
    elif sum(x) > (len(x) / 2):
        return "1"
    else:
        return "0"

def part_one():
    for i in range(0, len(data[0])):
        bitslist = list()
        for j in data:
            bitslist.append(int(j[i]))
        bit = popular_bit(bitslist)
        if bit == "both":
            print("wat.")
        elif bit == "0":
            gamma.append("0")
            epsilon.append("1")
        else:
            gamma.append("1")
            epsilon.append("0")
    gam = "".join(gamma)
    ep = "".join(epsilon)
    res = int(gam, 2) * int(ep, 2)
    return res

print("Part one:", part_one())

def part_two(data):
    for j in range(0, len(data[0])):
        temp_data = list()
        bitslist = list()
        for i in data:
            bitslist.append(int(i[j]))
        bit = popular_bit(bitslist)
        if bit == "both":
            for i in data:
                if i[j] == "1":
                    temp_data.append(i)
        else:
            for i in data:
                if i[j] == bit:
                    temp_data.append(i)
        data = [x for x in temp_data]
        if len(data) == 1:
            return int(data[0], 2)

def part_two_co(data):
    for j in range(0, len(data[0])):
        temp_data = list()
        bitslist = list()
        for i in data:
            bitslist.append(int(i[j]))
        bit = popular_bit(bitslist)
        if bit == "both":
            for i in data:
                if i[j] == "0":
                    temp_data.append(i)
        else:
            for i in data:
                if i[j] != bit:
                    temp_data.append(i)
        data = [x for x in temp_data]
        if len(data) == 1:
            return int(data[0], 2)

print("Part two:", part_two(data) * part_two_co(data))